### ISSUE TITLE
iOS emulation fails to start

### ISSUE THREAD

**Issue by bob8490** [nspire-emus/firebird#37]
So I've got the desktop version running great.  Emulating an nSpire CX, with CAS v3.1 installed.

My boot1 is named "boot1.img", and my flash is named "flash.img".  When I transfer these files to my phone (or my iPad), I'm met with an ambiguous "couldn't start emulation" or "could not resume," depending on which I try to do.

Files are in the correct Documents folder, transferred with SSH.  Installed the .ipa with the latest version of AppCake.  Both devices are jailbroken.  Phone is running 9.0.2, and iPad running 7.1.2.

Anything I can try?

**Comment by adriweb** [comment]
Hmm. It looks like you've followed the steps to get it working as expected, so I'm not sure what's happening...
Just in case, how did you get the boot1 and how did you generate the flash.img file?

**Comment by bob8490** [comment]
I used polydumper 4 on my nSpire CX for the boot1, boot2, manuf, and diags, and generated flash.img with the desktop version using those files and a clean copy of 3.1CX from TI.  I then installed 3.1CAS using nlaunch.

**Comment by Vogtinator** [comment]
It should definitely work on 64bit devices.
@adriweb: As you're responsible for the darwin builds, can you check again that it builds everything with -march=armv7?

**Comment by bob8490** [comment]
Here's a full crash log. Looks like the same error that mstallone had.

http://pastebin.com/etQ20ziM

**Comment by Vogtinator** [comment]
It shows "(32-bit)" so ARMv8-A isn't the issue here.
Sadly I can't do much with that trace, can you get a dump of the code at the crashing location?
Hexdump or disasm doesn't matter.

**Comment by bob8490** [comment]
Never used hexdump- can you provide a little more detail about what I need to do?

I know my way around a terminal, so no need to ELI5; just a bit of explanation.

**Comment by Vogtinator** [comment]
I have no idea how to debug iOS applications. @adriweb ?

**Comment by bob8490** [comment]
The syslog may include more information: http://pastebin.com/z3tmMw4S

I did try disabling all of my other tweaks to rule out any conflicts, but the behavior did not change.

**Comment by adriweb** [comment]
Hm, that's a lot of MobileSubstrate plugins. Could you try respringing in Safe Mode to see if Firebird likes it better?

**Comment by Vogtinator** [comment]
"MS:Error: recompile with -Wl,-segalign,4000" leads to http://www.idownloadblog.com/2015/10/15/cydia-substrate-0-9-6010/
So iOS 9.0 changed something that requires .dylibs to be recompiled. The sections in the trace look correctly aligned though...

**Comment by bob8490** [comment]
I did. No change.
Unfortunately, Crash Reporter doesn't work in safe mode, so no crash log. But the behavior was the same.

**Comment by bob8490** [comment]
I tried it on 7.1.2 as well on my iPad Air, with the same result.

**Comment by bob8490** [comment]
So I don't think it's because of 9.x

**Comment by bob8490** [comment]
Update:
I had forgotten to change the permissions and the owner for the img's on my iPad.
So it works fine on 7.1.2 on my iPad Air.

**Comment by bob8490** [comment]
More info: I disabled all of my plugins except for crash reporter, and was left with this syslog:

Fri Feb 19 14:49:31 2016: firebird (com.apple.console): MS:Notice: Injecting: com.firebird.firebird [firebird] (1240.10)
Fri Feb 19 14:49:36 2016: ReportCrash (Crash Reporter): Formulating report for corpse[2662] firebird
Fri Feb 19 14:50:08 2016: firebird (com.apple.console): MS:Notice: Injecting: com.firebird.firebird [firebird] (1240.10)
Fri Feb 19 14:50:15 2016: ReportCrash (Crash Reporter): Formulating report for corpse[2696] firebird


The crash log was the same as before.

**Comment by adriweb** [comment]
Quick question, if you try to transfer the files with iTunes (with the app file sharing thing), does it work on the iOS 9 device?

**Comment by bob8490** [comment]
No :/
I believe that the iTunes file sharing feature makes mobile the owner, which is what allows firebird to read it.  Transferring to the iPad with iTunes will work straight away, transferring to iPad with SSH and then making mobile the owner will also work.
Neither work on my phone, though.

**Comment by bob8490** [comment]
How do I compile this with Xcode?  I'd like to tinker with it a bit, but I've never actually used Xcode itself, just for g++ and such.

**Comment by adriweb** [comment]
Well actually you'd need the Qt SDK first, which you can get for free on their website. Then use Qt Creator with the readme instructions.
Also, read this: https://github.com/nspire-emus/firebird#a-special-case-ios-with-translation-jit-enabled - however there's another way:
Once built (and failed), open the .xcodeproj file in the build folder, and open the `asmcode_arm.S` file and check the checkbox on the right sidebar to add the file to the app target - it will then get assembled and linked, and you just have to hit "run" to launch it on your device (you may need ), including with debug support.

I'll try to build the app with the stack alignment sometime soon if you don't get it running before me :)

**Comment by bob8490** [comment]
I believe I already have Qt from a project I did a few months back.. Thanks for the heads up.

I'm installing El Capitan right now so that I can get the Xcode update that supports iOS9.

I'll give all that a shot in about 45 minutes.

**Comment by adriweb** [comment]
Alright, note that Qt 5.4 or 5.6beta is recommended. There are issues with 5.5.

**Comment by bob8490** [comment]
Noted. Thank you.

**Comment by bob8490** [comment]
I have the project open in Qt creator, but I'm getting the error ":-1: error: You have to run 'git submodule init' and 'git submodule update' first."

Neither of those commands will run properly.  I get "fatal: Not a git repository (or any of the parent directories): .git"

I assume this means I'm either not in the correct directory, or need additional setup before running these commands.

Sorry for my n00byness

**Comment by Vogtinator** [comment]
You need to "git clone git://github.com/nspire-emus/firebird.git" and use that directory.

**Comment by bob8490** [comment]
regarding asmcode_arm.o;
/Release-iphoneos  does not exist inside of [BUILD_FOLDER]/firebird.build/.

Should I create that folder (and subfolders)?

**Comment by adriweb** [comment]
Hmm, you're getting close. The build folder (and subfolders) only get created when a build is done - have you done that yet?
If using the manual instructions (not Qt Creator), it may just be inside the `bulid` folder.

**Comment by bob8490** [comment]
I'm using Qt creator.

inside /[BUILD LOCATION]/firebird.build/,  all I have is Debug-iphoneos and Debug-iphonesimulator.

**Comment by adriweb** [comment]
Oh yeah, you're building in Debug then, that's fine (won't be as fast as Release). You can choose Release in the bottom left of Qt Creator with the little arrow or something similar next to the run/debug/buttons.

**Comment by bob8490** [comment]
Got it.

So now, inside of Release-iphoneos/firebird.build/, I have a file called dgph and that's it..

Should I create "Objects-normal/armv7/"?

**Comment by adriweb** [comment]
Hm, nope, all the .o files should be there somewhere, when you build. I suppose you can look at the build log so you can see the path.

**Comment by bob8490** [comment]
They do exist inside of the Debug-iphonesimulator directory..

Can you / would you mind getting on Skype to sort this out?  Would be much easier with screensharing if you have the time.

**Comment by adriweb** [comment]
Oh but that just means you built for the simulator and not for an actual connected device (see https://doc.qt.io/qtcreator/creator-developing-ios.html).

I don't think I'd have anything more to show you than what I have:
![release iphoneos](https://i.imgur.com/OVAu7Zn.png)

With that, you'll see the proper files :)
(And you can also delete the simulator target in the Projects tab)

**Comment by bob8490** [comment]
So. Many. Errors.  Every time I make progress, more errors!

Now I get this when I try to build with Xcode:

clang: warning: argument unused during compilation: '-mno-thumb'
error: unknown target CPU 'armv7-a'
Command /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang failed with exit code 1

**Comment by bob8490** [comment]
Yeah, I've got just random little discrepancies from what I'm supposed to be seeing..

**Comment by adriweb** [comment]
>error: unknown target CPU 'armv7-a'

Oh yeah, that stupid error, I forgot every time that it needs to be changed >.> Good news is that it looks like you're almost there lol
So, in Xcode, go here in the build settings :
![https://i.imgur.com/uBAPdnL.png](xcode build settings)
and replace all instances of `armv7-a` by `armv7`

I'm also trying the build with the latest commits, see if I get same issues as you...

**Comment by bob8490** [comment]
Now it says 'armv7' is unknown.  Wtf, xcode.

**Comment by adriweb** [comment]
Ugh, actually, I just opened the .xcodeproj file, did NOT change the armv7-a stuff, but added `-stdlib=libc++` in the "Other C Flags", and it compiles and runs fine on my iPhone...

**Comment by adriweb** [comment]
Do you have the proper iOS SDK on Xcode btw ? (Look at the available updates in Xcode preferences)

**Comment by bob8490** [comment]
Didn't work for me :/

What kind of phone / what OS are you using?

I didn't have the simulators, but I don't see a download for the SDK, nor do I have any available updates.  I'm downloading the simulators now just in case.

**Comment by adriweb** [comment]
Well, the latest version of Xcode should be enough, it has the required stuff. No actual need for additional simulators.
I'm using an iPhone 5, iOS 9.0.2 JB.

I'll try giving you an .ipa with the stack alignement thing

**Comment by bob8490** [comment]
Okay.
Can you attempt to compile for ios9.0 and arm64?

**Comment by adriweb** [comment]
Well, there's no arm64 JIT in firebird, so I'm not sure that's possible.
A 64-bit build without JIT at all would however probably be possible. I'll try both

**Comment by bob8490** [comment]
64 without JIT probably has the best shot at working, I would think.

**Comment by adriweb** [comment]
Yep, but that's going to be quite slow...

**Comment by bob8490** [comment]
True.. But it may(?) be tolerable on the 6.. And at least it would work haha

**Comment by adriweb** [comment]
This thing should be arm64. Untested...

[firebird_iOS_arm64_noJIT_untested.zip](https://github.com/nspire-emus/firebird/files/139397/firebird_iOS_arm64_noJIT_untested.zip)

**Comment by bob8490** [comment]
App won't even open :/

**Comment by adriweb** [comment]
Yeah I'm not really surprised, I had never tried an arm64, maybe the Qt stuff don't actually support it.
Anyway, it's with your Xcode setup that things need to be tried anyway :( It should be very close... If you do a clean build, what/where are the errors again?

**Comment by bob8490** [comment]
Would be so much easier with a screenshare..

I think you may have been right before about it having to do with iOS9, because the iPad is also arm64, but it works fine with 7.1.2.

The error was that 'armv7' is an unknown target CPU.  I also tried with Cortex-A8 and arm64.  Same error both times.

**Comment by adriweb** [comment]
Here's a standard noJIT build, with the stack alignement option, it's worth trying...
[firebird.ipa.zip](https://github.com/nspire-emus/firebird/files/139401/firebird.ipa.zip)

**Comment by adriweb** [comment]
Well, I can't screenshare/skype/hangouts right now :/

**Comment by bob8490** [comment]
Okay, it's fine.

Trying that now.

**Comment by bob8490** [comment]
http://evilmcustoms.com/share/IMG_2848.PNG

Well that's odd.

Still doesn't start up :/  "Couldn't start emulation."

**Comment by bob8490** [comment]
This seems like it's a less serious error, though - it doesn't even try (presumably because it doesn't know the file names?).

**Comment by adriweb** [comment]
Nah, it looks worse, the stack alignment probably messed things up. So we're pretty much back at the standard ipa of the repo release page.

And considering it worked for another device of yours, I have no idea what's wrong. I guess we'll need to put more details in the error message if possible, to see what's going on...



Thanks,
Adrien Bertrand

**Comment by bob8490** [comment]
Ah :/

Well, hopefully I can jump in at a future release.

I'll poke at it for a couple days and see what I can do on my own with Google.

I really appreciate your willingness to spend time working on this. Thank you.

**Comment by adriweb** [comment]
I'll try to get you an .ipa with debug popup windows at "each" step, to see where stuff went wrong with the emulation starting.

**Comment by bob8490** [comment]
👌
You're awesome.

**Comment by adriweb** [comment]
Hopefully later today... I was pretty busy :P

**Comment by bob8490** [comment]
No problem. Standing by.

Sent from my iPhone

On Apr 24, 2016, at 2:02 PM, Adrien Bertrand <notifications@github.com<mailto:notifications@github.com>> wrote:


Hopefully later today... I was pretty busy :P

-

**Comment by adriweb** [comment]
So, here's a 32-bit* debug build with JIT turned off, to start simply... [firebird_dbg_32bit_nojit.ipa.zip](https://github.com/nspire-emus/firebird/files/233791/firebird_dbg_32bit_nojit.ipa.zip)
_(* well, just standard settings, nothing forced for 64bit)_

1. Remove any version you may have installed already
2. Install this .ipa (it's signed with my dev key, but since you're jailbroken, I suppose it's not a problem?)
3. Transfer `flash.img` and `boot1.img` with iTunes with the File Sharing thing in the apps sync panel
4. Launch the app on the device
5. Slide the keyboard to the left and select both boot1 and flash images from the app's document folder (there's a shortcut on the left of the file chooser)
6. If it works, well, great, but if it doesn't for some reason, copy/paste the content of the log file the app will have created in the documents folder (visible from iTunes' file sharing thing)

I've also just fixed (146995d) HiDPI-related issues (as far as I can tell) that made the LCD appear twice as big on retina iOS devices (at least with Qt 5.6 builds...)

**Comment by bob8490** [comment]
You got it.
Not home at the moment, will do this when I get back.

The dev key won't be an issue.

FYI, the reason you have to use iTunes file sharing is that the files must have mobile as the owner and be RWX by anyone.  I did it manually with chmod on my iPad.

Anyway, will shoot you an update this evening.

**Comment by adriweb** [comment]
Yeah if you have files transferred another way with manually adjusted permissions it should be fine.
However, my phone isn't jailbroken anymore so I'm testing things the official way now :P

**Comment by bob8490** [comment]
before emu_start
do_resume = 0
   ERROR in emu_start (not resuming) (couldnt flash_open)
success == false in EmuThread::run !
before emu_start
do_resume = 1
Trying to resume from snapshot...
   ERROR while resuming from snapshot (snapshot_size < sizeof(emu_snapshot))
success == false in EmuThread::run !
before emu_start
do_resume = 0
   ERROR in emu_start (not resuming) (couldnt flash_open)
success == false in EmuThread::run !
QKqueueFileSystemWatcherEngine::addPaths: open: Operation not permitted
QKqueueFileSystemWatcherEngine::addPaths: open: Operation not permitted
QKqueueFileSystemWatcherEngine::addPaths: open: Operation not permitted
QKqueueFileSystemWatcherEngine::addPaths: open: Operation not permitted

**Comment by bob8490** [comment]
All files are owned by mobile, and RWX by all users.

I did not use iTunes file transfer, as Firebird couldn't access its own Data folder. (in iOS9, the Data folder and Application bundle are in separate directories)  I instead created a 'Documents' folder inside the .app folder, and then set permissions.

**Comment by adriweb** [comment]
Hm, how did you install the .ipa? Using AppSync or similar, it should be installed properly enough such that iTunes sees it and can transfers files etc...
But still, with a 777 chmod, it should have worked...

Anyway, here's a build with even more debugging things, it might help (perror logged on the fopen): [firebird_dbg_32bit_nojit_moredebug.ipa.zip](https://github.com/nspire-emus/firebird/files/233875/firebird_dbg_32bit_nojit_moredebug.ipa.zip)

**Comment by bob8490** [comment]
AppCake, which uses AppSync.

iTunes could see the Data folder, but the app itself couldn't.. When I go to browse for the boot1 and flash images, I can only view the .app folder in the browser.

I'll give that one a shot.  Give me 10 mins.

**Comment by adriweb** [comment]
On the file browser when I want to choose the boot1/flash, the second item from the list on the left is a shortcut to "Documents" (the app's) - do you have that as well?
When I transfer the files with iTunes, it's where they end up.

**Comment by bob8490** [comment]
HOLY SHNIKEYS, BATMAN.

Apparently it MUST be in that directory.
I do have that shortcut; I had overlooked it.

With the files in that directory and the permissions set, it worked a treat.
I'd bet money it has something to do with sandbox... smh.

**Comment by Vogtinator** [comment]
Does the JIT work fine as well?

**Comment by bob8490** [comment]
How do I determine that?

**Comment by Vogtinator** [comment]
Well, are you using the JIT or non-JIT build?

**Comment by bob8490** [comment]
The one you sent me is JIT, yes? The "moredebug" version is what I have installed at the moment.

**Comment by adriweb** [comment]
No, the ones attached in this thread aren't JIT (which is why they're probably rather slow).
I'll try to get you a JIT one soon.

**Comment by bob8490** [comment]
It is rather.  Usable though.

Thank you.  Much appreciated.

**Comment by ethanx94** [comment]
Having the same issue. While firebird_dbg_32bit_nojit_moredebug.ipa works I wonder how fast it would run with under a JIT version? ;-)

**Comment by adriweb** [comment]
Hm i realiwed I never uploaded the JIT build here, sorry :/
I don't have a jailbroken iOS device to test here myself, so I'll try to get something but I can't really promise it'll work.

**Comment by adriweb** [comment]
Good news, I'll have my hands on a 64bit iPhone in the next few days/weeks, so I'll be building (and testing!) a new version ASAP.

**Comment by bob8490** [comment]
Awesome!!
As long as you're poking at it, can we get a tablet version that lets us see the entire keypad? 😬 Even on my phone I'd be happy with a smaller keypad if I didn't have to scroll.

**Comment by adriweb** [comment]
Hmmm, well @Vogtinator said he was thinking about a whole mobile UI redesign, so maybe this can be looked into during/after that :)

**Comment by adriweb** [comment]
Alright so, I've been able to build and run Firebird for on an iPhone 7 / iOS 10. Obviously without JIT but even if it's slow (at boot especially ; the rest is bearable).
The whole process needed some fiddling because Qt doesn't generate a valid Xcode 8 project... I'll see if I can manage something more or less automatic.

Also, for posterity's sake... :P
> 12:39:34 <@Vogtinato> Yay
> 12:39:54 <@Adriweb> you'll just have to make a 64-bit arm jit
> 12:39:56 <@Vogtinato> I'll have access to a rpi3 soon, so I may be able to write a aarch64 JIT
> 12:39:56 — @Adriweb runs
> 12:40:06 <@Adriweb> ninjad
> 12:40:07 — @Vogtinato catches Adriweb
> 12:40:27 <@Vogtinato> The architectures are actually really really similiar
> 12:40:52 <@Vogtinato> As I had to make the core 64bit ready for AMD64 anyway, it's practically just replacing the actually emitted code
> 12:41:19 <@Vogtinato> (Except for the add,mov,sub,etc., which needs to be handled separately...)

**Comment by adriweb** [comment]
A new version has been released: https://github.com/nspire-emus/firebird/releases/latest
Please try it and see if that fixes the problem(s).

Note that for now, I've been unable to get a full arm64 stable build, as there are some yet-unexplained crashes.
The current build (it's 32-bit) has been tested successfully on an iPhone 7 (so, arm64), since the JIT disables itself at runtime if necessary.

Closing the issue. Re-open if needed.

**Comment by adriweb** [comment]
Great news: we've managed to get the JIT working on iOS, even when non-jailbroken, and on both 32-bit and 64-bit devices!
This will be in the 1.2 version, probably released tomorrow :)
